### PR TITLE
patch only svc/svc_cx symbols

### PR DESCRIPTION
**Pull request by cbouvet-ledger**

### CODE REVIEW

File: speculos/main.py [hunk 1]
```diff
@@ -94,15 +94,17 @@ def get_elf_infos(app_path):
...
-        sym_code = symtab.get_symbol_by_name('_code')
-        code_start = sym_code[0]['st_value'] if sym_code else 0
-        sym_ecode = symtab.get_symbol_by_name('_ecode')
-        code_end = sym_ecode[0]['st_value'] if sym_ecode else 0
+        svc_call_addr = 0
+        svc_cx_call_addr = 0
+        if symtab.get_symbol_by_name("SVC_Call"):
+            svc_call_addr = symtab.get_symbol_by_name("SVC_Call")[0]['st_value'] & (~1)
+        if symtab.get_symbol_by_name("SVC_cx_call"):
+            svc_cx_call_addr = symtab.get_symbol_by_name("SVC_cx_call")[0]['st_value'] & (~1)
```
Comment (fbeutin-ledger) [inline speculos/main.py:102]:
```suggestion
        svc_call_symbol = symtab.get_symbol_by_name("SVC_Call")
        if svc_call_symbol is not None:
            svc_call_addr = svc_call_symbol[0]['st_value'] & (~1)
        svc_cx_call_symbol = symtab.get_symbol_by_name("SVC_cx_call")
        if svc_cx_call_symbol is not None:
            svc_cx_call_addr = svc_cx_call_symbol[0]['st_value'] & (~1)
```

File: speculos/main.py [hunk 2]
```diff
@@ -94,15 +94,17 @@ def get_elf_infos(app_path):
             sys.exit(1)
         estack = sym_estack[0]['st_value']
 
-        sym_code = symtab.get_symbol_by_name('_code')
-        code_start = sym_code[0]['st_value'] if sym_code else 0
-        sym_ecode = symtab.get_symbol_by_name('_ecode')
-        code_end = sym_ecode[0]['st_value'] if sym_ecode else 0
+        svc_call_addr = 0
```
Comment (fbeutin-ledger) [inline speculos/main.py:97]:
```suggestion
        # Look for the symbols SVC_Call and SVC_cx_call
        # if they are found, save their addresses to patch them to replace the SYSCALL later
        svc_call_addr = 0
```

File: src/launcher.c [hunk 3]
```diff
@@ -171,22 +173,41 @@ int replace_current_code(struct app_s *app)
...
     _exit(1);
   }
 
-  if (patch_svc(memory.code + start, code_length) != 0) {
-    /* this should never happen, because the svc were already patched without
-     * error during the first load */
-    _exit(1);
+  uint32_t text_start = (hw_model != MODEL_NANO_S) ? TEXT_START : LNS_TEXT_START;
+
+  if (app->elf.svc_call_addr != 0 || app->elf.svc_cx_call_addr != 0) {
```
Comment (fbeutin-ledger) [inline src/launcher.c:183]:
```suggestion
  // If svc_call_addr or svc_cx_call_addr are defined (non 0),
  // patch the elf at this address to remove the SYSCALL
  if (app->elf.svc_call_addr != 0 || app->elf.svc_cx_call_addr != 0) {
```

File: speculos/main.py (hunk 1 above)
Comment (fbeutin-ledger) [reply speculos/main.py:102]: Only one parsing per symbol + `is not None` is more explicit and recommended

File: src/launcher.c (hunk 3 above)
Comment (fbeutin-ledger) [reply src/launcher.c:183]: Just comments

File: src/launcher.c [hunk 4]
```diff
@@ -171,22 +173,41 @@ int replace_current_code(struct app_s *app)
...
+      uint32_t code_length = 2;
+
+      if (patch_svc(memory.code + start, code_length) != 0) {
+        /* this should never happen, because the svc were already patched
+         * without error during the first load */
+        _exit(1);
+      }
+    }
+  } else {
+    if (patch_svc(memory.code, app->elf.load_size) != 0) {
```
[the remark this answers is not in the document]
Comment (cbouvet-ledger) [reply src/launcher.c:206]: :shrug: just making sure it's done

File: src/launcher.c [hunk 5]
```diff
@@ -315,21 +336,41 @@ static void *load_app(char *name)
...
 
-  if (patch_svc(code + start, code_length) != 0) {
-    warn("could not patch svcs in app code");
-    goto error;
+  uint32_t text_start = (hw_model != MODEL_NANO_S) ? TEXT_START : LNS_TEXT_START;
+
+  if (app->elf.svc_call_addr != 0 || app->elf.svc_cx_call_addr != 0) {
+    if (app->elf.svc_call_addr != 0) {
+      uint32_t start = app->elf.svc_call_addr - text_start;
+      uint32_t code_length = 3;
```
[the remark this answers is not in the document]
Comment (cbouvet-ledger) [reply src/launcher.c]: fixed patch_svc

**Review by abonnaudet-ledger — APPROVED** [review]

**Pull request closed without merging**